File: apps/mobile/plugins/remote/android/src/main/java/com/velaterm/remote/TrustPromptCoordinator.kt
package com.velaterm.remote

import java.util.UUID

/** All methods and presenter callbacks run on the UI executor supplied by the owner. */
internal class TrustPromptCoordinator(
    private val schedule: (Long, () -> Unit) -> (() -> Unit),
    private val presenter: (Request, (Boolean) -> Unit) -> (() -> Unit),
    private val timeoutMillis: Long = 90_000,
    private val nowMillis: () -> Long = { System.nanoTime() / 1_000_000 },
) {
    data class Request(val identity: String, val fingerprint: String, val changed: Boolean)
    private data class Waiter(val id: UUID, val current: () -> Boolean, val complete: (Boolean) -> Unit)
    private class Entry(val request: Request, val deadline: Long) {
        val waiters = mutableListOf<Waiter>()
        var dismiss: (() -> Unit)? = null
        var timeout: (() -> Unit)? = null
        var finished = false
    }
    private val queue = mutableListOf<Entry>()
    private var active: Entry? = null

    fun decide(request: Request, current: () -> Boolean, id: UUID = UUID.randomUUID(), complete: (Boolean) -> Unit) {
        if (!current()) { complete(false); return }
        val entry = (listOfNotNull(active) + queue).firstOrNull {
            !it.finished && it.request.identity == request.identity && it.request.fingerprint == request.fingerprint
        } ?: Entry(request, nowMillis() + timeoutMillis).also {
            queue.add(it)
            // Joining callers share this original deadline, including time spent waiting in the queue.
            it.timeout = schedule(timeoutMillis) { finish(it, false) }
        }
        entry.waiters.add(Waiter(id, current, complete))
        if (nowMillis() >= entry.deadline) { finish(entry, false); return }
        schedule(0) { startNext() }
    }
    fun invalidate() {
        for (entry in queue.toList() + listOfNotNull(active)) {
            if (nowMillis() >= entry.deadline) { finish(entry, false); continue }
            val stale = entry.waiters.filter { !it.current() }
            entry.waiters.removeAll(stale.toSet())
            stale.forEach { it.complete(false) }
            if (entry.waiters.isEmpty()) finish(entry, false)
        }
    }
    fun cancel(id: UUID) {
        for (entry in queue.toList() + listOfNotNull(active)) {
            val waiter = entry.waiters.firstOrNull { it.id == id } ?: continue
            entry.waiters.remove(waiter); waiter.complete(false)
            if (entry.waiters.isEmpty()) finish(entry, false)
            return
        }
    }
    fun close() {
        val pending = queue.toList(); queue.clear()
        for (entry in pending + listOfNotNull(active)) finish(entry, false)
    }
    private fun startNext() {
        if (active != null) return
        while (queue.isNotEmpty()) {
            val entry = queue.removeAt(0)
            if (entry.finished) continue
            active = entry; invalidate()
            if (entry.finished) continue
            try {
                val dismiss = presenter(entry.request) { finish(entry, it) }
                if (entry.finished) dismiss() else entry.dismiss = dismiss
            } catch (_: Exception) { finish(entry, false) }
            return
        }
    }
    private fun finish(entry: Entry, decision: Boolean) {
        if (entry.finished) return
        val accepted = decision && nowMillis() < entry.deadline
        entry.finished = true
        entry.timeout?.invoke(); entry.timeout = null
        try { entry.dismiss?.invoke() } catch (_: Exception) { } finally { entry.dismiss = null }
        val waiters = entry.waiters.toList(); entry.waiters.clear()
        queue.remove(entry)
        if (active === entry) active = null
        for (waiter in waiters) waiter.complete(accepted && waiter.current())
        schedule(0) { startNext() }
    }
}

/** Thread-safe load identity is checked again while holding the lock used for committing trust. */
internal class TrustLoadState {
    private var generation = 0
    private var closed = false
    private val decisions = mutableMapOf<Pair<String, String>, Boolean>()
    @Synchronized fun begin(): Int { generation++; closed = false; decisions.clear(); return generation }
    @Synchronized fun close() { generation++; closed = true; decisions.clear() }
    @Synchronized fun epoch(): Int = generation
    @Synchronized fun isCurrent(epoch: Int) = !closed && generation == epoch
    @Synchronized fun decision(identity: String, fingerprint: String): Boolean? = decisions[identity to fingerprint]
    @Synchronized fun record(epoch: Int, identity: String, fingerprint: String, decision: Boolean) {
        if (isCurrent(epoch)) decisions[identity to fingerprint] = decision
    }
}

File: apps/mobile/plugins/remote/ios/Sources/VelaRemotePlugin/TrustPromptCoordinator.swift
import Foundation

/// Main-actor ownership keeps prompt completion, cancellation and invalidation atomic.
@MainActor final class TrustPromptCoordinator {
    struct Request: Hashable, Sendable {
        let identity: String
        let fingerprint: String
        let changed: Bool
    }
    /// The returned closure must dismiss this presentation and cancel any pending presentation work.
    typealias Presenter = @Sendable @MainActor (Request, @escaping @MainActor (Bool) -> Void) throws -> (() -> Void)
    private struct Waiter {
        let id: UUID
        let continuation: CheckedContinuation<Bool, Never>
        let isCurrent: () -> Bool
    }
    private final class Entry {
        let request: Request
        let deadline: ContinuousClock.Instant
        var waiters: [Waiter] = []
        var dismiss: (() -> Void)?
        var watchdog: Task<Void, Never>?
        var finished = false
        init(_ request: Request, deadline: ContinuousClock.Instant) { self.request = request; self.deadline = deadline }
    }
    private nonisolated let presenter: Presenter
    private nonisolated let timeout: TimeInterval
    private nonisolated let now: @MainActor @Sendable () -> ContinuousClock.Instant
    private var queue: [Entry] = []
    private var active: Entry?
    var isPresenting: Bool { active != nil }

    nonisolated init(timeout: TimeInterval = 90,
                     now: @escaping @MainActor @Sendable () -> ContinuousClock.Instant = { ContinuousClock.now },
                     presenter: @escaping Presenter) {
        self.timeout = timeout; self.now = now; self.presenter = presenter
    }

    func decide(_ request: Request, isCurrent: @escaping () -> Bool = { true }) async -> Bool {
        guard !Task.isCancelled, isCurrent() else { return false }
        let id = UUID()
        return await withTaskCancellationHandler(operation: {
            await withCheckedContinuation { continuation in
                let matching = ([active].compactMap { $0 } + queue).first {
                    !$0.finished && $0.request.identity == request.identity && $0.request.fingerprint == request.fingerprint
                }
                let entry = matching ?? Entry(request, deadline: now().advanced(by: .seconds(timeout)))
                entry.waiters.append(Waiter(id: id, continuation: continuation, isCurrent: isCurrent))
                if matching == nil { queue.append(entry); watch(entry) }
                if now() >= entry.deadline { finish(entry, false); return }
                Task { self.startNext() }
            }
        }, onCancel: { Task { @MainActor in self.cancel(id) } })
    }

    /// Called before a connection or load can be replaced; the watchdog also handles external invalidation.
    func cancelInvalidRequests() {
        for entry in queue + [active].compactMap({ $0 }) {
            if now() >= entry.deadline { finish(entry, false); continue }
            let stale = entry.waiters.filter { !$0.isCurrent() }
            entry.waiters.removeAll { !$0.isCurrent() }
            stale.forEach { $0.continuation.resume(returning: false) }
            if entry.waiters.isEmpty { finish(entry, false) }
        }
    }
    func cancelAll() {
        let pending = queue; queue.removeAll()
        for entry in pending { finish(entry, false) }
        if let active { finish(active, false) }
    }
    private func cancel(_ id: UUID) {
        for entry in queue + [active].compactMap({ $0 }) {
            if let index = entry.waiters.firstIndex(where: { $0.id == id }) {
                entry.waiters.remove(at: index).continuation.resume(returning: false)
                if entry.waiters.isEmpty { finish(entry, false) }
                return
            }
        }
    }
    private func startNext() {
        guard active == nil else { return }
        while !queue.isEmpty {
            let entry = queue.removeFirst()
            guard !entry.finished else { continue }
            active = entry
            cancelInvalidRequests()
            guard !entry.finished else { continue }
            do {
                let dismiss = try presenter(entry.request) { [weak self, weak entry] decision in
                    guard let self, let entry else { return }; self.finish(entry, decision)
                }
                // A presenter may fail or answer synchronously before it returns its dismissal handle.
                if entry.finished { dismiss() } else { entry.dismiss = dismiss }
            } catch { finish(entry, false) }
            return
        }
    }
    // Queueing and presentation share the first caller's monotonic deadline; joining never renews it.
    private func watch(_ entry: Entry) {
        entry.watchdog = Task { [weak self, weak entry] in
            while let self, let entry, !entry.finished {
                self.cancelInvalidRequests()
                do { try await Task.sleep(nanoseconds: 20_000_000) } catch { return }
            }
        }
    }
    private func finish(_ entry: Entry, _ decision: Bool) {
        guard !entry.finished else { return }
        let accepted = decision && now() < entry.deadline
        entry.finished = true; entry.watchdog?.cancel(); entry.watchdog = nil
        entry.dismiss?(); entry.dismiss = nil
        let waiters = entry.waiters; entry.waiters.removeAll()
        queue.removeAll { $0 === entry }
        if active === entry { active = nil }
        for waiter in waiters { waiter.continuation.resume(returning: accepted && waiter.isCurrent()) }
        Task { self.startNext() }
    }
}

/// Decisions belong to an explicit connection load, never to a later retry or another browser.
@MainActor final class TrustLoadState {
    private(set) var generation = 0
    private(set) var closed = false
    private var decisions: [String: [String: Bool]] = [:]
    func begin() { generation += 1; closed = false; decisions.removeAll() }
    func close() { generation += 1; closed = true; decisions.removeAll() }
    func isCurrent(_ epoch: Int) -> Bool { !closed && epoch == generation }
    func decision(_ identity: String, _ fingerprint: String) -> Bool? { decisions[identity]?[fingerprint] }
    func record(_ decision: Bool, identity: String, fingerprint: String, epoch: Int) {
        guard isCurrent(epoch) else { return }
        decisions[identity, default: [:]][fingerprint] = decision
    }
}
